libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -621,58 +621,121 @@ void ContractLevelChecker::checkBaseABICompatibility(ContractDefinition const& _
 
 void ContractLevelChecker::checkAmbiguousOverrides(ContractDefinition const& _contract) const
 {
-	vector<FunctionDefinition const*> contractFuncs = _contract.definedFunctions();
-
-	auto const resolvedBases = resolveDirectBaseContracts(_contract);
-
-	FunctionMultiSet inheritedFuncs = inheritedFunctions(&_contract);;
+	// Fetch non overridden base functions and sort them by signature.
+	FunctionMultiSet nonOverriddenBaseFunctionSet = convertContainer<FunctionMultiSet>(
+		nonOverriddenBaseFunctions(&_contract)
+	);
+	// Remove all functions that match the signature of a function in the current contract.
+	nonOverriddenBaseFunctionSet -= _contract.definedFunctions();
 
-	// Check the sets of the most-inherited functions
-	for (auto it = inheritedFuncs.cbegin(); it != inheritedFuncs.cend(); it = inheritedFuncs.upper_bound(*it))
+	// Walk through the set of functions signature by signature.
+	for (auto it = nonOverriddenBaseFunctionSet.cbegin(); it != nonOverriddenBaseFunctionSet.cend();)
 	{
-		auto [begin,end] = inheritedFuncs.equal_range(*it);
+		auto nextSignature = nonOverriddenBaseFunctionSet.upper_bound(*it);
+		std::set baseFunctions(it, nextSignature, [](auto const* a, auto const* b) { return a->id() < b->id(); });
+		it = nextSignature;
 
-		// Only one function
-		if (next(begin) == end)
-			continue;
-
-		// Not an overridable function
-		if ((*it)->isConstructor())
+		if (baseFunctions.size() > 1)
 		{
-			for (begin++; begin != end; begin++)
-				solAssert((*begin)->isConstructor(), "All functions in range expected to be constructors!");
-			continue;
-		}
+			struct OverrideGraph {
+				std::map<FunctionDefinition const*, int> nodes;
+				std::map<int, FunctionDefinition const*> nodeInv;
+				std::map<int, std::set<int>> edges;
+				int numNodes = 2;
+				void addEdge(int a, int b)
+				{
+					edges[a].insert(b);
+					edges[b].insert(a);
+				}
+				int visit(FunctionDefinition const* function)
+				{
+					auto it = nodes.find(function);
+					if (it != nodes.end())
+						return it->second;
+					int currentNode = numNodes++;
+					nodes[function] = currentNode;
+					nodeInv[currentNode] = function;
+
+					if (!function->overrides())
+						addEdge(currentNode, 1);
+					else
+						for (auto const* baseFunction: function->annotation().baseFunctions)
+							addEdge(currentNode, visit(baseFunction));
 
-		// Function has been explicitly overridden
-		if (contains_if(
-			contractFuncs,
-			[&] (FunctionDefinition const* _f) {
-				return hasEqualNameAndParameters(*_f, **it);
-			}
-		))
-			continue;
+					return currentNode;
```

Comment:
Why not use AST IDs?

Reply:
And what value would I use for the artificial vertex? For the root I could use the current contract... ok - I could use ``-1u`` for the artificial one, but that's also hacky? This way I get contiguous IDs and don't have to use sets and maps in the ``CutVertexFinder``s...